apps/src/turtle/turtle.js

```diff
@@ -840,32 +871,28 @@ Artist.prototype.animate = function() {
   this.smoothAnimateStepSize = (stepSpeed === 0 ?
     FAST_SMOOTH_ANIMATE_STEP_SIZE : SMOOTH_ANIMATE_STEP_SIZE);
 
-  if (this.level.editCode) {
-    var stepped = true;
-    while (stepped) {
-      codegen.selectCurrentCode(this.interpreter,
-                                this.cumulativeLength,
-                                this.userCodeStartOffset,
-                                this.userCodeLength,
-                                this.studioApp_.editor);
-      try {
-        stepped = this.interpreter.step();
-      }
-      catch(err) {
-        // TODO (cpirich): populate lineNumber as we do for studio/applab:
-        this.executionError = { err: err, lineNumber: 1 };
-        this.finishExecution_();
-        return;
-      }
-      stepped = this.interpreter.step();
+  if (this.level.editCode &&
+      this.JSInterpreter &&
+      this.JSInterpreter.initialized()) {
+      
+    var programDone = false;
+    var completedTuple = false;
 
-      if (this.executeTuple_()) {
-        // We stepped far enough that we executed a commmand, break out:
-        break;
+    do {
+      programDone = this.JSInterpreter.isProgramDone();
+
+      if (!programDone) {
+        this.JSInterpreter.executeInterpreter();
+
+        completedTuple = this.executeTuple_();
```

Comment:
I might be worth adding a comment about why we need to `executeTuple_()` after each `executeInterpreter()` step.  I think because executing the interpreter only adds items to the tuple queue?